packages/hardhat-core/src/internal/artifacts/index.ts

```diff
@@ -0,0 +1,333 @@
+import {
+  Artifact,
+  Artifacts as IArtifacts,
+  ArtifactSource,
+  BuildInfo,
+  CompilerInput,
+  CompilerOutput,
+} from "../../types";
+
+import { ARTIFACT_FORMAT_VERSION } from "../constants";
+import { HardhatError } from "../core/errors";
+import { ERRORS } from "../core/errors-list";
+
+import { ReadOnlySource } from "./readonly";
+import { MutableSource } from "./mutable";
+import { CachingSource } from "./caching";
+
+type SupportedArtifactSource =
+  | ArtifactSource
+  | ReadOnlySource
+  | MutableSource
+  | CachingSource;
+
+export class Artifacts implements IArtifacts {
+  constructor(
+    artifactsPath: string,
+    private readonly _sourcesInPriorityOrder: SupportedArtifactSource[] = [
+      new CachingSource(artifactsPath),
+    ]
+  ) {}
+
+  public async readArtifact(
+    contractNameOrFullyQualifiedName: string
+  ): Promise<Artifact> {
+    const artifact = await this._getFirstValueFromSources(
+      "readArtifact",
+      contractNameOrFullyQualifiedName
+    );
+    if (artifact === undefined) {
+      throw new HardhatError(ERRORS.ARTIFACTS.NOT_FOUND, {
+        contractName: contractNameOrFullyQualifiedName,
+        suggestion: "",
+      });
+    }
+    return artifact;
+  }
+
+  public readArtifactSync(contractNameOrFullyQualifiedName: string): Artifact {
+    const artifact = this._getFirstValueFromSourcesSync(
+      "readArtifactSync",
+      contractNameOrFullyQualifiedName
+    );
+    if (artifact === undefined) {
+      throw new HardhatError(ERRORS.ARTIFACTS.NOT_FOUND, {
+        contractName: contractNameOrFullyQualifiedName,
+        suggestion: "",
+      });
+    }
+    return artifact;
+  }
+
+  public artifactExists(
+    contractNameOrFullyQualifiedName: string
+  ): Promise<boolean> {
+    return this._getFirstValueFromSources(
+      "artifactExists",
+      contractNameOrFullyQualifiedName
+    );
```

Comment:
I think this is a bug. The first source will never return `undefined`, so its response will always be used. Instead, this should check all sources and return `true` if any returns `true`, and `false` otherwise.

Btw, I was on the fence about the `_getFirstValueFromSources` approach, but this convinced me that we shouldn't do it. This feels more like incidental duplication than something we should abstract over.

Reply:
Addressed in https://github.com/NomicFoundation/hardhat/pull/3285/commits/3647b0444ccb9f122dba864575f15de2b24f1032